sql/event_log: reduce tech debt, simplify and fix bugs

knz opened the pull request:
Fixes #64685.
First commit from #64871. 

Prior to this patch, there was a little mess in `sql/event_log.go`
that had been introduced when "optimizing" GRANT/REVOKE to only
use one write batch for all the events logged:

- a lot of code had been duplicated;
- the control flow had been rendered more complex;
- the API interface for the functions in event_log.go
  were complex to use, requiring callers to provide descriptor
  IDs and structured events as separate slices;
- the optimizing logic was not properly applied to the
  other case where multiple events are emitted:
  SQL audit logging in `exec_log.go`.

This patch streamlines this by reducing `event_log.go` back to its
simpler form: an overall event refinement pipeline with a
straightforward control flow.

To guide future maintainers, the patch also adds an explanatory
comment at the top of the file that sketches the overall structure of
the pipeline.

Finally, this patch fixes a bug introduced when query logging
started using structured events: the ability to automatically
copy all the execution events to the DEV channel when
setting the `vmodule` setting to `exec_log=2` (or above).

In addition to fixing that bug, the following new vmodule-based
abilities are added:

- events for DDL statements and others that call `logEvent()` can now
  be collected in the DEV channel by using the name of the source file
  where they were generated as filter (e.g. `vmodule=create_table=2` for
  the CREATE TABLE events.
- events of other kinds can be collected in the DEV channel
  by setting `vmodule=event_log=2`.

(Note a subtle difference between `vmodule=create_table=2` and
`vmodule=exec_log=2`: the former emits the event to the DEV channel
while the stmt is executed; the latter emits the event after the stmt
completes. If both are enabled, TWO events are sent to the DEV channel.)

Since all the vmodule filtering options are subject to change without
notice between versions, we do not wish to document these nuances.
For this reason, the release note below is left blank.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/64894)
<!-- Reviewable:end -->

rafiss reviewed (commented):
so far i wasn't able to follow the code enough to identify where the fix to log the query string in exec_log happens. in general, the comments and refactors look good to me.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64894#-:-M_WzVhb6TEUoa_eSc7T:btdvll5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RichardJCai)



<!-- Sent from Reviewable.io -->

knz commented:
> so far i wasn't able to follow the code enough to identify where the fix to log the query string in exec_log happens

it was a mistake to do `if logV { log.VEventf(...) }` in `exec_log.go` because at that point the structured payload is not fully populated yet. This population occurs in the pipeline described at the top of `event_log.go`. 

So instead the `log.V` case is handled differently: `if logStmt || logV { logEvent(...) }` and then deeper inside the pipeline (inside `insertEventRecords()`) we check log.V again to do the thing at that point.

Does this explain?

knz committed:
sql: simplify the internal event logging APIs

Prior to this patch, there was a little mess in `sql/event_log.go`
that had been introduced when "optimizing" GRANT/REVOKE to only
use one write batch for all the events logged:

- the API interface for the functions in event_log.go
  were complex to use, requiring callers to provide descriptor
  IDs and structured events as separate slices;
- the optimizing logic was not properly applied to the
  other case where multiple events are emitted:
  SQL audit logging in `exec_log.go`.

This patch cleans this up by using the same struct `eventLogEntry` as
argument to the various APIs.

Release note: None

knz committed:
sql: use a single code path for both single-event and multi-event logging

The previous patch to batch event writes for GRANT/REVOKE had
duplicated code. This was not necessary. This patch fixes this by
using the same code for both cases.

Release note: None

knz committed:
sql: rename the system event log cluster setting

This clarifies its purpose.

Release note: None

knz committed:
sql: move a function to its right place

`logEventsOnlyExternally()` is specific to the code in `exec_log.go`
and should thus reside there. This patch achieves that.

Release note: None

knz committed:
sql/event_log: avoid boolean parameters

The code was previously using two booleans `onlyLog` and
`writeToEventLog` which were making the code difficult to understand
and to maintain.

This patch fixes this by introducing a bitset with descriptive names.

Release note: None

knz committed:
sql: pass SQL exec details as a struct to the event log functions

This further simplifies the internal API.

Release note: None

knz committed:
sql/event_log: clarify the API in an explanatory comment

This patch describes `event_log.go` at a high level: an overall event
refinement pipeline with a straightforward control flow.

Release note: None

knz committed:
sql/event_log: fix bug with verbose logging

This patch fixes a bug introduced when query logging
started using structured events: the ability to automatically
copy all the execution events to the DEV channel when
setting the `vmodule` setting to `exec_log=2` (or above).

In addition to fixing that bug, the following new vmodule-based
abilities are added:

- events for DDL statements and others that call `logEvent()` can now
  be collected in the DEV channel by using the name of the source file
  where they were generated as filter (e.g. `vmodule=create_table=2` for
  the CREATE TABLE events.
- events of other kinds can be collected in the DEV channel
  by setting `vmodule=event_log=2`.

(Note a subtle difference between `vmodule=create_table=2` and
`vmodule=exec_log=2`: the former emits the event to the DEV channel
while the stmt is executed; the latter emits the event after the stmt
completes. If both are enabled, TWO events are sent to the DEV channel.)

Since all the vmodule filtering options are subject to change without
notice between versions, we do not wish to document these nuances.
For this reason, the release note below is left blank.

Release note: None

knz reviewed (commented):
I have rewritten the PR to decompose each logical change into a sub-commit, for your reviewing pleasure. I hope you like it?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64894#-:-M_fFpdLEPj-iBQO1hmA:b9e34su)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RichardJCai)



<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
sorry i missed your updates while i was out of office at the end of last week. yes, this does help and it was pleasurable to review! lgtm

> Reviewed 1 of 4 files at r1, 5 of 5 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/64894#-:-MaCoQf637v-Y0y7cYvv:b-hdbibl)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RichardJCai)



<!-- Sent from Reviewable.io -->

knz commented:
TFYR!

bors r=rafiss
